core-library/src/main/java/org/silverpeas/core/contribution/publication/dao/DistributionTreeCriteria.java

```diff
@@ -0,0 +1,122 @@
+ * License, or (at your option) any later version.
+ *
+ * As a special exception to the terms and conditions of version 3.0 of
+ * the GPL, you may redistribute this Program in connection with Free/Libre
+ * Open Source Software ("FLOSS") applications as described in Silverpeas's
+ * FLOSS exception.  You should have received a copy of the text describing
+ * the FLOSS exception, and it is also available here:
+ * "https://www.silverpeas.org/legal/floss_exception.html"
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.silverpeas.core.contribution.publication.dao;
+
+import org.silverpeas.core.util.StringUtil;
+
+import java.util.Collection;
+import java.util.Optional;
+
+import static java.util.Collections.emptyList;
+import static java.util.Optional.ofNullable;
+
+/**
+ * Class that allows to set distribution tree criteria for publications and nodes.
+ * @author silveryocha
+ */
+class DistributionTreeCriteria {
+  private final String instanceId;
+  private boolean checkVisibility = false;
+  private String statusSubQuery = null;
+  private Collection<String> instanceIdsToIgnore = null;
+
+  /**
+   * Initializes distribution tree criteria axed on the given component instance id.
+   * @param instanceId identifier of the aimed component instance host.
+   * @return an initialized {@link DistributionTreeCriteria} instance.
+   */
+  static DistributionTreeCriteria onInstanceId(final String instanceId) {
+    return new DistributionTreeCriteria( instanceId);
+  }
+
+  /**
+   * Hidden constructor
+   */
+  private DistributionTreeCriteria(final String instanceId) {
+    this.instanceId = instanceId;
+  }
+
+  /**
+   * Sets the visibility check.
+   * @param checkVisibility true to verify the visibility, false otherwise.
+   * @return itself.
+   */
+  DistributionTreeCriteria withVisibilityCheck(final boolean checkVisibility) {
+    this.checkVisibility = checkVisibility;
+    return this;
+  }
+
+  /**
+   * Sets manually the filtering on the status.
+   * @param statusSubQuery a string representing the SQL status filtering part.
+   * @return itself.
+   */
+  DistributionTreeCriteria withManualStatusFilter(final String statusSubQuery) {
+    this.statusSubQuery = statusSubQuery;
+    return this;
+  }
+
+  /**
+   * Sets manually the filtering on the status.
+   * @param instanceIdsToIgnore a collection of component instance identifier which MUST be ignored.
+   * @return itself.
+   */
+  DistributionTreeCriteria ignoringInstanceIds(final Collection<String> instanceIdsToIgnore) {
+    this.instanceIdsToIgnore = instanceIdsToIgnore;
```

Comment:
Hum... You shouldn't expect the caller won't modify externally the collection it passes in this method. To avoid any possible pb, you should copy the collection instead of just keep its reference.

Reply:
It is not necessary here to take more memory from my point of view.
In addition to that, this criteria class is scoped to the DAO package and its use is exclusively internal to it.

Reply:
I have finally refactored the criteria so that it can be used from anywhere. Concerned service signature has been modificied to use it as parameter.
I took into account your feedback about the collection copy.